### ISSUE TITLE
Metis is required right now for Gunrock installation but it shouldn't be...

### ISSUE THREAD

**Issue by neoblizz** [gunrock/gunrock#97]
`find_package(Metis 5.0 REQUIRED)` this doesn't give an error/warning in CMakeLists, I will have it corrected in issue #76 where it will warn the users. 

But after talking to @sgpyc it is not suppose to be required, Gunrock should be able to work as long as it doesn't use Metis. Support can be added where Metis `includes` can be commented for use without this dependency.

**Comment by jowens** [comment]
oops

**Comment by jowens** [comment]
I support fixing this.

**Comment by neoblizz** [comment]
@sgpyc (or @yzhwang) I would like to try this, can you point me to the places where metis is included (or if there is a way I can search that on my own)?

**Comment by sgpyc** [comment]
It's here https://github.com/gunrock/gunrock/blob/dev/gunrock/app/metisp/metis_partitioner.cuh#L17

**Comment by neoblizz** [comment]
Ok, so this is really cool. I worked on this issue @sgpyc and here's my fix for not forcing users to include/use `metis_partitioner.cuh`:
Edits to `CMakeLists.txt` so that it adds a pre-compiler flag `METIS_FOUND` that can be checked in the `metis_partitioner` file using `add_definitions`:

```
FIND_LIBRARY(METIS_LIB NAMES metis)
find_package(Metis ${gunrock_REQUIRED_METIS_VERSION} REQUIRED)
SET(METIS_FOUND FALSE)
IF (METIS_LIB)
  SET(METIS_FOUND TRUE)
  add_definitions( -DMETIS_FOUND=${METIS_FOUND}) #Adds compiler definition METIS_FOUND = TRUE
  message(STATUS "Found Metis")
ELSE (METIS_LIB)
  add_definitions( -DMETIS_FOUND=${METIS_FOUND}) #Adds compiler definition METIS_FOUND = FALSE
  message(WARNING "Metis was requested but support was not found, run /dep/install_metis.sh for installation.")
ENDIF (METIS_LIB)
```

Then, just add `#ifdef METIS_FOUND ... #endif` in the  `metis_partitioner` file. Now, here's the cool thing that I didn't know about! You can actually output pre-compiler definitions using the following code, during `make` it tells me if Metis was found by checking the `METIS_FOUND` and prints `TRUE` or `FALSE`:
```
#define XSTR(x) STR(x)
#define STR(x) #x

#pragma message "Metis Found? " XSTR(METIS_FOUND)
```
Thoughts if I should implement this? @sgpyc @yzhwang @jowens

**Comment by neoblizz** [comment]
How the test build looks like when printing `METIS_FOUND` (in this build I forced set it to false):

```
[  1%] Building NVCC (Device) object gunrock/CMakeFiles/gunrock.dir/__/externals/moderngpu/src/gunrock_generated_mgpucontext.cu.o
[  2%] Building NVCC (Device) object gunrock/CMakeFiles/gunrock.dir/app/bfs/gunrock_generated_bfs_app.cu.o
/home/muhammad/gunrock/gunrock/app/metisp/metis_partitioner.cuh:18:33: note: #pragma message: Metis Found? FALSE
 #pragma message "Metis Found? " XSTR(METIS_FOUND)
                                 ^
[  3%] Building NVCC (Device) object gunrock/CMakeFiles/gunrock.dir/app/bc/gunrock_generated_bc_app.cu.o
/home/muhammad/gunrock/gunrock/app/metisp/metis_partitioner.cuh:18:33: note: #pragma message: Metis Found? FALSE
 #pragma message "Metis Found? " XSTR(METIS_FOUND)
                                 ^
[  4%] Building NVCC (Device) object gunrock/CMakeFiles/gunrock.dir/app/cc/gunrock_generated_cc_app.cu.o
/home/muhammad/gunrock/gunrock/app/metisp/metis_partitioner.cuh:18:33: note: #pragma message: Metis Found? FALSE
 #pragma message "Metis Found? " XSTR(METIS_FOUND)
                                 ^
[  5%] Building NVCC (Device) object gunrock/CMakeFiles/gunrock.dir/app/pr/gunrock_generated_pr_app.cu.o
/home/muhammad/gunrock/gunrock/app/metisp/metis_partitioner.cuh:18:33: note: #pragma message: Metis Found? FALSE
 #pragma message "Metis Found? " XSTR(METIS_FOUND)
                                 ^
[  6%] Building NVCC (Device) object gunrock/CMakeFiles/gunrock.dir/app/sssp/gunrock_generated_sssp_app.cu.o
/home/muhammad/gunrock/gunrock/app/metisp/metis_partitioner.cuh:18:33: note: #pragma message: Metis Found? FALSE
 #pragma message "Metis Found? " XSTR(METIS_FOUND)
                                 ^
[  7%] Building NVCC (Device) object gunrock/CMakeFiles/gunrock.dir/util/gunrock_generated_test_utils.cu.o
[  8%] Building NVCC (Device) object gunrock/CMakeFiles/gunrock.dir/util/gunrock_generated_error_utils.cu.o
[  9%] Building NVCC (Device) object gunrock/CMakeFiles/gunrock.dir/util/gunrock_generated_misc_utils.cu.o
Scanning dependencies of target gunrock
```

**Comment by neoblizz** [comment]
Fixed in issue #138 (Temporary fix added for Makefiles, but will automatically work once we fix CMakeLists.txt issue #139).
